### PR TITLE
rfc: allow universe unification to solve `max u v =?= max u ?v`

### PR DESCRIPTION
**Pull request by kim-em**
See issue description at #2297.

### CODE REVIEW

**Comment by tydeu** [comment]
@semorrison Should not this be symmetric? That is, if Lean solves `max u v =?= max u ?v` should it not also solve `max u v =?= max ?u v`?

**Comment by semorrison** [comment]
@tydeu, my understanding is that `max`s have already been sorted by the time they arrive here, in a way that ensures the mvars don't appear in the first slot.

**Comment by tydeu** [comment]
@semorrison Ah, okay, makes sense. 👍

File: src/Lean/Meta/LevelDefEq.lean [hunk 1]
```diff
@@ -81,6 +81,11 @@ mutual
         match (← Meta.decLevel? v) with
         | some v => Bool.toLBool <$> isLevelDefEqAux u v
         | none   => return LBool.undef
+    -- Handle `max u ?_ =?= max u v` and `max u ?_ =?= max v u` for `u` a parameter.
+    | Level.max p@(.param _) m@(.mvar _), Level.max u v =>
+      if u == p then solve m v
+      else if v == p then solve m u
+      else return LBool.undef
```
Comment (semorrison) [inline src/Lean/Meta/LevelDefEq.lean:88]:
```suggestion
    | Level.max p@(.param _) (.mvar m), Level.max u v =>
      if u == p then
        assignLevelMVar m v
        return LBool.true
      else if v == p then
        assignLevelMVar m u
        return LBool.true
      else return LBool.undef
```
also works here, and is perhaps more conservative as there's no recursive call to `solve`, but I think should have the same behaviour. It also compiles mathlib4 fine, and fixes our problems there.

**Comment by leodemoura** [comment]
> @tydeu, my understanding is that `max`s have already been sorted by the time they arrive here, in a way that ensures the mvars don't appear in the first slot.

@semorrison You must confirm that. You also need to properly document `LevelDefEq.lean` and add a Note (in this module) about the  change that has been made.

**Comment by semorrison** [comment]
I'm going to label this as low priority for now, in the context of the `UnivLE` workaround I developed in Mathlib seeming to work for now. (See https://github.com/leanprover-community/mathlib4/pull/5723, https://github.com/leanprover-community/mathlib4/pull/5724, https://github.com/leanprover-community/mathlib4/pull/5726.) I don't think `UnivLE` is a complete solution, so this may be worth revisiting later.

One unfortunate aspect of the current set up is that universe unification issues as in the title often appear in error messages that obscure the "actual" error that the user needs to address (usually an instance not being available). We might think about how to improve that situation without changing the unification algorithm, although at present I don't have ideas in that direction. Anyone coming to this issue in the meantime who has been tripped up by the error messages: creating a separate issue about the error message situation would be a great intermediate step.

**Comment by semorrison** [comment]
Superseded by https://github.com/leanprover/lean4/pull/3981

**Pull request closed without merging**